unit/goto-programs/goto_program_validate.cpp

```diff
@@ -0,0 +1,429 @@
+  return true;
+}
+
+bool test_for_failure(
+  goto_modelt &goto_model,
+  const goto_model_validation_optionst goto_model_validation_options)
+{
+  bool caught{false};
+  try
+  {
+    validate(
+      goto_model.goto_functions,
+      validation_modet::EXCEPTION,
+      goto_model_validation_options);
+  }
+  catch(incorrect_goto_program_exceptiont &e)
+  {
+    caught = true;
+  }
+  return caught;
+}
+} // namespace
+
+SCENARIO("validate goto program")
+{
+  goto_modelt goto_model;
+
+  // void f(){int x = 1;}
+  symbolt x;
+  x.name = "x";
+  x.mode = ID_C;
+  x.type = signed_int_type();
+  goto_model.symbol_table.add(x);
+
+  symbolt f;
+  f.name = "f";
+  f.mode = ID_C;
+  f.type = code_typet({}, empty_typet()); // void rtn, take no params
+
+  code_blockt f_body;
+  f_body.copy_to_operands(code_declt(x.symbol_expr()));
+  f_body.copy_to_operands(
+    code_assignt(x.symbol_expr(), from_integer(1, signed_int_type())));
+  f.value = f_body;
+  goto_model.symbol_table.add(f);
+
+  // void g(){int y = 2;}
+  symbolt y;
+  y.name = "y";
+  y.mode = ID_C;
+  y.type = signed_int_type();
+  goto_model.symbol_table.add(y);
+
+  symbolt g;
+  g.name = "g";
+  g.mode = ID_C;
+  g.type = code_typet({}, empty_typet());
+
+  code_blockt g_body;
+  g_body.copy_to_operands(code_declt(y.symbol_expr()));
+  g_body.copy_to_operands(
+    code_assignt(y.symbol_expr(), from_integer(2, signed_int_type())));
+  g.value = g_body;
+  goto_model.symbol_table.add(g);
+
+  symbolt z;
+  z.name = "z";
+  z.mode = ID_C;
+  z.type = signed_int_type();
+  goto_model.symbol_table.add(z);
+
+  symbolt fn_ptr;
+  fn_ptr.name = "fn_ptr";
+  fn_ptr.mode = ID_C;
+  // pointer to fn call
+  fn_ptr.type = pointer_typet(code_typet{{}, empty_typet()}, 64);
+  goto_model.symbol_table.add(fn_ptr);
+
+  symbolt entry_point;
+  entry_point.name = goto_functionst::entry_point();
+  entry_point.mode = ID_C;
+  entry_point.type = code_typet({}, void_typet());
+
+  code_blockt entry_point_body;
+  entry_point_body.copy_to_operands(code_declt(z.symbol_expr()));
+  entry_point_body.copy_to_operands(
+    code_assignt(z.symbol_expr(), from_integer(3, signed_int_type())));
+
+  entry_point_body.copy_to_operands(
+    code_assignt(fn_ptr.symbol_expr(), address_of_exprt{f.symbol_expr()}));
+  entry_point_body.copy_to_operands(make_void_call(g.symbol_expr()));
+
+  entry_point.value = entry_point_body;
+  goto_model.symbol_table.add(entry_point);
+
+  /// check entry_point_exists()
+  WHEN("the goto program has no entry point")
+  {
+    goto_convert(goto_model, null_message_handler);
+
+    auto &function_map = goto_model.goto_functions.function_map;
+    auto it = function_map.find(goto_functionst::entry_point());
+    function_map.erase(it);
+
+    THEN("fail!")
+    {
+      goto_model_validation_optionst validation_options;
+      validation_options.disable_all_checks();
+      validation_options.entry_point_exists = true;
+      REQUIRE(test_for_failure(goto_model, validation_options));
```

Comment:
As mentioned above, use the built-in `REQUIRE_THROWS_AS` macro instead, ditto for other uses below.

Reply:
Done.